csharp/src/Ice/OutgoingResponseFrame.cs

```diff
@@ -172,15 +176,22 @@ public OutgoingResponseFrame(Current current, RemoteException exception)
             IsSealed = true;
         }
 
-        private OutgoingResponseFrame(Encoding encoding, bool writeVoidReturnValue = false)
+        private OutgoingResponseFrame(Protocol protocol, Encoding encoding, bool writeVoidReturnValue = false)
         {
+            Protocol = protocol;
             Encoding = encoding;
             Data = new List<ArraySegment<byte>>();
             if (writeVoidReturnValue)
             {
-                Encoding.CheckSupported();
-                // TODO: this is not correct, the encoding of the empty payload must be `encoding`.
-                Data.Add(Ice1Definitions.EmptyResponsePayload);
+                // TODO: use encoding instead of the protocol's encoding for the empty encaps?
```

Comment:
I would be good to fix this TODO. To me using the protocol sounds more appropriate here given that the frame depends first on the protocol. The encapsulation payload might depend on the encoding next but right now, we don't have different empty payload encapsulation based on the encoding.

Reply:
Fixed with:
```
            if (writeVoidReturnValue)
            {
                if (Protocol == Protocol.Ice1 && Encoding == Encoding.V1_1)
                {
                    Data.Add(Ice1Definitions.EmptyResponsePayload);
                }
                else if (Protocol == Protocol.Ice2 && Encoding == Encoding.V2_0)
                {
                    Data.Add(Ice2Definitions.EmptyResponsePayload);
                }
                else
                {
                    var ostr = new OutputStream(Protocol.GetEncoding(), Data, new OutputStream.Position(0, 0));
                    ostr.WriteByte((byte)ReplyStatus.OK);
                    _ = ostr.WriteEmptyEncapsulation(Encoding);
                }
                Size = Data.GetByteCount();
                IsSealed = true;
            }
```

We use the IceXDefinitions bytes for the common ice2 + 2.0 and ice1 + 1.1 situations (where we could/should maintain static frames, see other comment). Otherwise, we create an OutputStream to fill this frame. This is what happen is your proxy looks like `foo -p ice2 -e 1.1` and you call an operation that returns void on this proxy. The client code expects a void return value encoded with 1.1, even through the ice2 frame header uses the 2.0 encoding.

Reply:
An alternative would be to allow the header's encoding in this void-return situation. When I call a void-returning operation on `foo -p ice2 -e 1.1`, it's ok to get a void-return encoded with the protocol's encoding (2.0) even though the request params were encoded with 1.1.

Reply:
> An alternative would be to allow the header's encoding in this void-return situation. When I call a void-returning operation on `foo -p ice2 -e 1.1`, it's ok to get a void-return encoded with the protocol's encoding (2.0) even though the request params were encoded with 1.1.

I think this would not be correct in a protocol bridging situation such as:

```
(3.7 client) -- ice1/1.1 --> (4.0 "router") -- ice2/2.0 --> (4.0 server)
```

Reply:
In which use case this code will be used? Shouldn't we simply reject the creation of empty responses for unknown protocol/encoding pairs?

Reply:
I think my example above is a valid use-case unless we disallow protocol bridging.

Reply:
Ok right, we're going to need few tests to ensure these combinations work fine 😄